### ISSUE TITLE
Report implicit overrides doesn't catch override through inheritance

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#4821]
```python
class  A:
    def f(self, x: int) -> None:
        pass


class  B:
    def f(self, x: int) -> None:
        pass


class  C(A, B):
    pass
```
In `C`, `A.f` overrides `B.f`, but no implicit override is reported.

I agree that this is bad code (not my code), but I think it should be reported as it's almost always an error.  This is incidentally another reason I think that `__init__` should not be reported as an implicit override.

**Comment by erictraut** [comment]
PEP 698 does not indicate as such. It says:

> We believe that @override is most useful if checkers also allow developers to opt into a strict mode where methods that override a parent class are required to use the decorator.

I don't have a strong opinion on this. If you would like to pursue a clarification from the PEP's authors on this point and there is consensus that overrides of peer classes in a multi-inheritance case should also be checked for implicit overrides, I'm willing to add it to pyright.

**Comment by erictraut** [comment]
Link to [python/typing discussion](https://github.com/python/typing/issues/1376).

**Comment by erictraut** [comment]
Thinking about this more, I don't think it makes sense to enforce the `@override` decorator in this case. An author of a class would have no reason to anticipate that someone might use their class in a multi-inheritance situation, so it's unreasonable to ask them to mark a method as `@override` to accommodate such a case. Furthermore, the `@override` would be inappropriate when the class is _not_ used in a multi-inheritance case.

**Comment by NeilGirdhar** [comment]
> An author of a class would have no reason to anticipate that someone might use their class in a multi-inheritance situation, so it's unreasonable to ask them to mark…

I think you've misunderstood what I'm asking for.  I'm not saying that the author of `A` or `B` needs to anticipate anything.  Those classes are perfectly fine without annotations.

What I'm asking for is for a typing error to be raised when `C` is defined, which creates an implicit override situation.  This is nearly always indicative of a bug.  Like you say, "the authors of A and B have no reason to anticipate that someone might use their class in multiple inheritance".  So when `C` is defined, it occludes `B.f`.  That means that any of `B`'s private methods that call `self.f` will get `A.f` instead!

**Comment by erictraut** [comment]
OK, so your request has nothing to do with PEP 698. I misunderstood because you used the phrase "implicit overload".

If you enable `reportIncompatibleMethodOverrides`. Pyright already emits an error if `A` and `B` both define a method of the same name with different (incompatible) signatures. 

I think it's legitimate (and not necessarily an indication of a bug) if `A` and `B` define methods of the same name that have compatible signatures, so it wouldn't make sense to flag that as an error.

**Comment by NeilGirdhar** [comment]
> If you enable `reportIncompatibleMethodOverride`. Pyright already emits an error if `A` and `B` both define a method of the same name with different (incompatible) signatures.

Yes, I understand that.

> I think it's legitimate (and not necessarily an indication of a bug) if `A` and `B` define methods of the same name that have compatible signatures, so it wouldn't make sense to flag that as an error.

I don't think this would ever be good design.  Can you come up with a situation in which this would be reasonable 
 Normally, if `A` and `B` both implement the same method (not just in the sense of same signature, but in the sense of behavior), then both `A` and `B` should inherit from an interface that defines that method.

The reason I think this should fall under PEP 698 is because it is an example of an "implicit override": `A.f` _overrides_ `B.f` in `C`.

**Comment by erictraut** [comment]
I've seen this pattern used in a legitimate manner. It shouldn't be considered an error just because you consider it "not good design". (Personally, I think all multiple inheritance is "bad design", but I'm not going to change pyright to emit an error when someone uses multiple inheritance.)

If you want to enforce that rule in your code base, that's your prerogative, but it sounds like something that would be more appropriately handled by a linter, not a type checker. After all, it is not a type violation.

**Comment by NeilGirdhar** [comment]
> it sounds like something that would be more appropriately handled by a linter, not a type checker. After all, it is not a type violation.

Nothing caught by the strict mode of implicit overrides is a "type violation".  Why don't we wait to see what people thing in the python-typing thread?
